agent/util-scripts/pbench-tool-meister-start

```diff
@@ -244,13 +202,621 @@ def waitpid(pid):
+                    pass
+            except Exception:
+                logger.exception("failed to create localhost tool meister, daemonized")
+                failures += 1
+                tms[host] = {"pid": None, "status": "failed"}
+            else:
+                # Record the child pid to wait below.
+                tms[host] = {"pid": pid, "status": "forked"}
+        else:
+            args[0] = host
+            args[4] = tm_param_key
+            ssh_args = base_args + args
+            logger.debug(
+                "5b. starting remote tool meister, ssh_path=%r ssh_args=%r",
+                ssh_path,
+                ssh_args,
+            )
+            try:
+                pid = os.spawnv(os.P_NOWAIT, ssh_path, ssh_args)
+            except Exception:
+                logger.exception(
+                    "failed to create a tool meister instance for host %s", host
+                )
+                tms[host] = {"pid": None, "status": "failed"}
+            else:
+                # Record the child pid to wait below.
+                tms[host] = {"pid": pid, "status": "spawned"}
+
+    for host, tm_proc in tms.items():
+        if tm_proc["status"] == "failed":
+            failures += 1
+            continue
+        pid = tm_proc["pid"]
+        try:
+            exit_status = _waitpid(pid)
+        except Exception:
+            failures += 1
+            logger.exception(
+                "failed to create a tool meister instance for host %s", host
+            )
+        else:
+            if exit_status != 0:
+                failures += 1
+                logger.error(
+                    "failed to start tool meister on remote host '%s'"
+                    " (pid %d), exit status: %d",
+                    host,
+                    pid,
+                    exit_status,
+                )
+            else:
+                successes += 1
+
+    if failures > 0:
+        raise StartTmsErr(
+            "failures encountered creating tool miesters", ReturnCode.TMFAILURES
+        )
+    if successes == 0:
+        raise StartTmsErr(
+            "unable to successfully start any tool meisters,"
+            " but encountered no failures either",
+            ReturnCode.TMNOSUCCESSES,
+        )
+    if successes != tm_count:
+        raise StartTmsErr(
+            f"number of created Tool Meisters, {successes}, does not"
+            f" match the expected number of Tool Meisters, {tm_count}",
+            ReturnCode.TMMISSING,
+        )
+
+
+class BaseServer:
+    """BaseServer - abstract base class for common code shared between the
+    ToolDataSink and RedisServer classes.
+    """
+
+    envi_var = None
+    def_port = None
+    bad_port_code = None
+
+    class Err(ReturnCodeErr):
+        pass
+
+    def __init__(self):
+        # The pid of the managed server
+        self.host = None
+        self.port = None
+        self.bind_host = None
+        self.bind_port = None
+        self._managed = None
```

Comment:
What is the type of `self.managed`?

Reply:
From the two examples, the answer is `boolean`.

Reply:
Yes.  We start off a `None` indicating it is not set.  Notice that the assertions that use it assert `True`, not a boolean result value.